### PR TITLE
Drop headers

### PR DESCRIPTION
**Pull request by jakirkham**
Convert code from header files to Cython (generated C code is equivalent). Then drop the headers. These were there primarily for Windows compatibility with old Python versions (namely 2.7), but shouldn't be needed any more.

xref: https://github.com/zarr-developers/numcodecs/pull/368
xref: https://github.com/zarr-developers/numcodecs/pull/369

TODO:

- [ ] Unit tests and/or doctests in docstrings
- [ ] `tox -e py310` passes locally
- [ ] Docstrings and API docs for any new/modified user-facing classes and functions
- [ ] Changes documented in docs/release.rst
- [ ] `tox -e docs` passes locally
- [ ] GitHub Actions CI passes
- [ ] Test coverage to 100% (Coveralls passes)

### CODE REVIEW

**Comment by jakirkham** [comment]
@DimitriPapadopoulos, Idk how familiar you are with Cython/C, but would appreciate your review here (if you have time 🙂)

**Comment by DimitriPapadopoulos** [comment]
I'll have a look, but I actually don't know much about Cython. I am more experienced in Python bindings for C++ libraries using [SIP](https://www.riverbankcomputing.com/software/sip/intro).

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,8 @@ __pycache__/
 *$py.class
 
 # C extensions
+numcodecs/**/*.c
+numcodecs/**/*.h
```
Comment (DimitriPapadopoulos) [inline .gitignore:8]: Won't C files be generated by Cython? Shouldn't the resulting files be ignored by version control?

Comment (DimitriPapadopoulos) [inline .gitignore:8]: Had forgotten that. Thanks!

**Comment by DimitriPapadopoulos** [comment]
I think I get it:
* Most of the contents of `stdint_compat.h` are moved to `_utils.pxd` by d3250aebe3b98c3fbf65a860ec241712559a1657.
* The now (almost) empty `stdint_compat.h` is dropped in 2fb1d412a204e9dee801f7ca0a498ac77f1439b6.

File: numcodecs/_utils.pxd [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+# cython: profile=False
+# cython: linetrace=False
+# cython: binding=False
+# cython: language_level=3
+
+
+from libc.stdint cimport uint8_t, uint32_t
+
+
+cdef inline void store_le32(uint8_t c[4], uint32_t i) nogil:
```
Comment (DimitriPapadopoulos) [inline numcodecs/_utils.pxd:11]: Adding [nogil](url) is good.

**Review by DimitriPapadopoulos — APPROVED** [review]

File: .gitignore (hunk 1 above)
Comment (jakirkham) [reply .gitignore:8]: No worries. Did as well. Plus being able to filter the headers meant needing to drop the one that was here.

File: numcodecs/_utils.pxd (hunk 2 above)
Comment (jakirkham) [reply numcodecs/_utils.pxd:11]: Yep this forces Cython to generate straight C code. So the resulting code is nearly identical to what was in the header.

**Comment by jakirkham** [comment]
Yeah the dropping Python 2.7 & C files in your other two PRs reminded me there was some leftover compatibility code in this header for Windows Python 2.7. So this cleans that out and consolidates things into Cython (with some readability/code quality improvements).

**Review by joshmoore — COMMENTED** [review]
No strong opinions from my side, so generally :+1:.

Probably my only question is whether or not there will be any impact on client code. E.g. if `stdint_compat.h` doesn't get re-generated, could anyone have been depending on it ❓

**Comment by jakirkham** [comment]
Thanks Josh! 🙏

Nope the header isn't installed. It is only used as part of the build process.

**Pull request merged**